### PR TITLE
Add class to get workspace history

### LINKED ISSUE #23395: Scrape Project recovery files to send with error reporter
The ErrorReporter currently only sends minimal information. An obvious source of extra information would be the project recovery folder. If the user elects to share identifiable information this should be located zipped and sent with the post request.

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
**Description of work.**

This updates the error reporter so that if the user decided to share information the recovery folder is compressed and sent to the server along with the error report. 

**Report to:** keith 

**To test:**
* Select erroreports.mantidproject.org as your url
* Copy the file below into your mantidBuild/bin directory and run it with `python server_test.py`
* All the tests should pass but you should get some errors about failing to upload the recovery files.
[server_test.py.zip](https://github.com/mantidproject/mantid/files/2392633/server_test.py.zip)

Fixes #23395 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/ErrorReporter/retrieve_recovery_files.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+from mantid.kernel import ConfigService
+import os
+import shutil
+import hashlib
+import datetime
+
+
+class RetrieveRecoveryFiles(object):
```
Comment (DanNixon) [inline scripts/ErrorReporter/retrieve_recovery_files.py]: Why does this need to be a class at all? A module of free functions would be much cleaner.

Comment (Mantid-Matthew) [reply scripts/ErrorReporter/retrieve_recovery_files.py]: Given all the methods are static it doesn't need to be a class at all. You're correct a module of free functions would be cleaner. I'll update.

**Review by keeeto — DISMISSED** [review]
This works with the current database and also with a soon to come version that accepts the recovery files. This is going to be very useful in future for diagnosing instabilities.

File: Framework/Kernel/src/ErrorReporter.cpp [hunk 2]
```diff
@@ -34,16 +33,25 @@ Logger g_log("ErrorReporter");
...
 ErrorReporter::ErrorReporter(std::string application,
                              Types::Core::time_duration upTime,
                              std::string exitCode, bool share, std::string name,
                              std::string email, std::string textBox)
+    : ErrorReporter(application, upTime, exitCode, share, name, email, textBox,
+                    "") {}
+
+ErrorReporter::ErrorReporter(std::string application,
+                             Types::Core::time_duration upTime,
+                             std::string exitCode, bool share, std::string name,
```
Comment (soininen) [inline Framework/Kernel/src/ErrorReporter.cpp]: If I understood this correctly, `share` could be `const`.

File: Framework/Kernel/test/ErrorReporterTest.h [hunk 3]
```diff
@@ -93,6 +93,69 @@ class ErrorReporterTest : public CxxTest::TestSuite {
     TS_ASSERT_EQUALS(root["exitCode"].asString(), "0");
     TS_ASSERT_EQUALS(root["name"].asString(), "name");
     TS_ASSERT_EQUALS(root["email"].asString(), "email");
+    TS_ASSERT_EQUALS(root["textBox"].asString(), "textBox");
+  }
+
+  void test_errorMessageWithShareAndRecoveryFileHash() {
+    std::string name = "My testing application name";
```
Comment (soininen) [inline Framework/Kernel/test/ErrorReporterTest.h]: `const` to all constants in the tests.

**Review by soininen — CHANGES_REQUESTED** [review]
Just a few tiny cases of constness.

**Review by soininen — APPROVED** [review]

**Pull request merged**